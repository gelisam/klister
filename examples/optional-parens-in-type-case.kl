#lang "prelude.kl"

(import "optional-parens.kl")
(import (shift "prelude.kl" 1))
(import (shift "optional-parens.kl" 1))

(define-macros
  ([m1
    -- all three types can appear in type-patterns with parentheses
    (lambda (stx)
      (>>= (which-problem)
        (lambda (problem)
          (case problem
            [(expression tp)
             (type-case tp
               [(Nat1)
                (pure '(zero1))]
               [(Nat2)
                (pure '(zero2))]
               [(Integer)
                (pure '42)]
               [(List3 elem-tp)
                (pure '(nil3))])]))))]
   [m2
    -- the nullary types can appear in type-patterns without parentheses
    (lambda (stx)
      (>>= (which-problem)
        (lambda (problem)
          (case problem
            [(expression tp)
             (type-case tp
               [Nat1
                (pure '(suc1 m1))]
               [Nat2
                (pure '(suc2 m1))]
               [Integer
                (pure '(+ m1 1))]
               [(List3 elem-tp)
                (pure '(cons3 m1 m1))])]))))]))

(example (the Nat1 m2))
(example (the Nat2 m2))
(example (the Integer m2))
(example (the (List3 Integer) m2))
