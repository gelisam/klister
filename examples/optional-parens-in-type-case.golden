(suc1 (zero1)) : Nat1
(suc2 (zero2)) : Nat2
43 : Integer
(cons3 42 (nil3)) : (List3 Integer)
